frontend/src/app/main/data/workspace/thumbnails.cljs

```diff
@@ -0,0 +1,174 @@
+  "Stream that will inform the frame-wrapper to mount into memory"
+  [id]
+  (->> st/stream
+       (rx/filter (ptk/type? ::force-render))
+       (rx/map deref)
+       (rx/filter #(= % id))
+       (rx/take 1)))
+
+(defn update-thumbnail
+  "Updates the thumbnail information for the given frame `id`"
+  [id data]
+  (let [lock (uuid/next)]
+    (ptk/reify ::update-thumbnail
+      IDeref
+      (-deref [_] {:id id :data data})
+      
+      ptk/UpdateEvent
+      (update [_ state]
+        (-> state
+            (assoc-in [:workspace-file :thumbnails id] data)
+            (cond-> (nil? (get-in state [::update-thumbnail-lock id]))
+              (assoc-in [::update-thumbnail-lock id] lock))))
+
+      ptk/WatchEvent
+      (watch [_ state stream]
+        (when (= lock (get-in state [::update-thumbnail-lock id]))
+          (let [stopper (->> stream (rx/filter (ptk/type? :app.main.data.workspace/finalize)))
+                params {:file-id (:current-file-id state)
+                        :object-id id}]
+            ;; Sends the first event and debounce the rest. Will only make one update once
+            ;; the 2 second debounce is finished
+            (rx/merge
+             (->> stream
+                  (rx/take-until stopper)
+                  (rx/filter (ptk/type? ::update-thumbnail))
+                  (rx/filter #(= id (:id (deref %))))
+                  (rx/debounce 2000)
+                  (rx/first)
+                  (rx/flat-map
+                   (fn [event]
+                     (let [data (:data @event)]
+                       (rp/mutation! :upsert-file-object-thumbnail (assoc params :data data)))))
+
+                  (rx/map #(fn [state] (d/dissoc-in state [::update-thumbnail-lock id]))))
+
+             (->> (rx/of (update-thumbnail id data))
+                  (rx/observe-on :async)))))))))
+
+(defn remove-thumbnail
+  [id]
+  (ptk/reify ::remove-thumbnail
+    ptk/UpdateEvent
+    (update [_ state]
+      (-> state (d/dissoc-in [:workspace-file :thumbnails id])))
+
+    ptk/WatchEvent
+    (watch [_ state _]
+      (let [params {:file-id (:current-file-id state)
+                    :object-id id
+                    :data nil}]
+        (->> (rp/mutation! :upsert-file-object-thumbnail params)
+             (rx/ignore))))))
+
+(defn- extract-frame-changes
+  "Process a changes set in a commit to extract the frames that are changing"
+  [[event [old-objects new-objects]]]
+  (let [changes (-> event deref :changes)
+
+        extract-ids
+        (fn [{type :type :as change}]
+          (case type
+            :add-obj [(:id change)]
+            :mod-obj [(:id change)]
+            :del-obj [(:id change)]
+            :reg-objects (:shapes change)
+            :mov-objects (:shapes change)
+            []))
+
+        get-frame-id
+        (fn [id]
+          (let [shape (or (get new-objects id)
+                          (get old-objects id))]
+
+            (or (and (= :frame (:type shape)) id)
```

Comment:
`cph/frame-shape?` can be used of direct shape type checking

Reply:
ok